File: solutions/_1046_last_stone_weight/solution.java
package _1046_last_stone_weight;
import java.util.*;

class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (int stone : stones) {
            pq.add(stone);
        }

        while (pq.size() > 1) {
            int y = pq.poll();
            int x = pq.poll();
            if (x != y) {
                pq.add(y - x);
            }
        }
        return pq.isEmpty() ? 0 : pq.poll();
    }
}
  
/**
 * https://leetcode.com/problems/last-stone-weight/
 */

File: solutions/_1046_last_stone_weight/solution.cpp
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

class Solution
{
public:
    int lastStoneWeight(vector<int> &stones)
    {
        priority_queue<int> pq;
        for (int stone : stones)
        {
            pq.push(stone);
        }

        while (pq.size() > 1)
        {
            int y = pq.top();
            pq.pop();
            int x = pq.top();
            pq.pop();
            if (x != y)
            {
                pq.push(y - x);
            }
        }
        return pq.empty() ? 0 : pq.top();
    }
};

/*
https://leetcode.com/problems/last-stone-weight/
*/